### ISSUE TITLE
Occasionally failing unit tests

### ISSUE THREAD

**Issue by jmezach** [neo4j/neo4j-dotnet-driver#77]
As previously discussed in #52 I see unit tests that are occasionally failing when I check out the 1.1 branch and repeatedly run the unit tests (from Visual Studio 2015).

The tests are failing with a _System.NullReferenceException_. The exceptions occur at different places but they all originate from the **Neo4j.Driver.Internal.Packstream.PackStream.Unpacker** class and are related to the _private static BitConverterBase _bitConverter_ variable. Here's the relevant code from that class:

``` c#
public class Unpacker
{
    private readonly IInputStream _in;
    private static BitConverterBase _bitConverter;

    public Unpacker(IInputStream inputStream, BitConverterBase converter)
    {
        _in = inputStream;
        _bitConverter = converter;
    }
}
```

This looks a bit weird since the constructor sets a static variable, not an instance variable. Then again this might be by design.

**Comment by zhenlineo** [comment]
I am thinking that you got the error probably because you were executing your tests concurrently. We got problems to run ITs concurrently against the server so we always run with `-parallel none` ([code](https://github.com/neo4j/neo4j-dotnet-driver/blob/1.1/Neo4j.Driver/RunUnitTests.bat))

I am not sure if we would fix the problem in unit tests as the ITs cannot run concurrently anyway (they need to clean the store after each run at least).

Let me know if you still get the problem if you run tests in a single thread.

**Comment by jmezach** [comment]
I tried to run the tests again without running them in parallel, but they still seem to fail sometime:

![image](https://cloud.githubusercontent.com/assets/1225489/15577206/ccbf5e2c-235b-11e6-86a2-c62a1c21efc6.png)

Running them in parallel doesn't seem to make the matter worse.

**Comment by zhenlineo** [comment]
ummm... may I ask what tool you are using to run these tests? I use reshaper and apparently there is no way to configure to not run in parallel. I would like to try the tool you are using.

**Comment by jmezach** [comment]
I'm using the built-in Visual Studio test runner. I only installed the xunit.runner.visualstudio NuGet package to get them to show up there.

**Comment by zhenlineo** [comment]
Okay, good to know such a tool.

So I could also reproduce the null exception locally with resharper unit test tool. I looked in the code where a test failed with null exception. However the failure does not make any sense to me as the test itself `new`s a bitConverter. Therefore I doubt the problem could be running concurrently where two tests try to new and set the static field for `Unpacker` objects at the same time.

Also for our test build, we use xunit command line tool to run the test with this [bat file](https://github.com/neo4j/neo4j-dotnet-driver/blob/1.1/Neo4j.Driver/RunUnitTests.bat), I do not remember I have ever saw this error on our test build server.

But if you still get the problem after you turned off parallel running, then maybe something else is wrong. I will continue investigating tomorrow. 

Meanwhile you could also try to change the static field to a normal instance to see if the problem goes away (If so, then it is still a concurrent execution problem, is't it!)

**Comment by jmezach** [comment]
I changed the static field to a normal instance variable and then ran the tests a couple of times and they don't seem to be failing anymore.

**Comment by zhenlineo** [comment]
It turns out to be a bug, the fix -> https://github.com/neo4j/neo4j-dotnet-driver/pull/80

Thanks so much to bring up this issue!
The PR would be merged next week after pair review to make sure I fixed it properly.

**Comment by zhenlineo** [comment]
Closing this issue after the fix #80 merged. Please be free to re-open this card if you still have the problem.
